cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll

```diff
@@ -372,185 +359,425 @@ cached private module Cached {
-        definitionReachesRank(vvar, block, defRank, reachesRank - 1) and
-        not hasDefinitionAtRank(vvar, block, reachesRank - 1, _)
+        definitionReachesRank(useLocation, block, defRank, reachesRank - 1) and
+        not hasDefinitionAtRank(useLocation, _, block, reachesRank - 1, _)
       )
     )
   }
 
   /**
-   * Holds if the definition of `vvar` at `(defBlock, defRank)` reaches the end of
-   * block `block`.
-   */
-  private predicate definitionReachesEndOfBlock(Alias::VirtualVariable vvar, OldBlock defBlock,
+    * Holds if the definition that overlaps `useLocation` at `(defBlock, defRank)` reaches the end of
+    * block `block` without any intervening definitions that overlap `useLocation`.
+    */
+  predicate definitionReachesEndOfBlock(Alias::MemoryLocation useLocation, OldBlock defBlock,
       int defRank, OldBlock block) {
-    hasDefinitionAtRank(vvar, defBlock, defRank, _) and
+    hasDefinitionAtRank(useLocation, _, defBlock, defRank, _) and
     (
       (
         // If we're looking at the def's own block, just see if it reaches the exit
         // rank of the block.
         block = defBlock and
-        variableLiveOnExitFromBlock(vvar, defBlock) and
-        definitionReachesRank(vvar, defBlock, defRank, exitRank(vvar, defBlock))
+        locationLiveOnExitFromBlock(useLocation, defBlock) and
+        definitionReachesRank(useLocation, defBlock, defRank, exitRank(useLocation, defBlock))
       ) or
       exists(OldBlock idom |
-        definitionReachesEndOfBlock(vvar, defBlock, defRank, idom) and
-        noDefinitionsSinceIDominator(vvar, idom, block)
+        definitionReachesEndOfBlock(useLocation, defBlock, defRank, idom) and
+        noDefinitionsSinceIDominator(useLocation, idom, block)
       )
     )
   }
 
   pragma[noinline]
-  private predicate noDefinitionsSinceIDominator(Alias::VirtualVariable vvar, OldBlock idom,
+  private predicate noDefinitionsSinceIDominator(Alias::MemoryLocation useLocation, OldBlock idom,
       OldBlock block) {
     Dominance::blockImmediatelyDominates(idom, block) and // It is sufficient to traverse the dominator graph, cf. discussion above.
-    variableLiveOnExitFromBlock(vvar, block) and
-    not hasDefinition(vvar, block, _)
+    locationLiveOnExitFromBlock(useLocation, block) and
+    not hasDefinition(useLocation, _, block, _)
   }
 
-  private predicate definitionReachesUseWithinBlock(Alias::VirtualVariable vvar, OldBlock defBlock,
-      int defRank, OldBlock useBlock, int useRank) {
-    defBlock = useBlock and
-    hasDefinitionAtRank(vvar, defBlock, defRank, _) and
-    hasUseAtRank(vvar, useBlock, useRank, _) and
-    definitionReachesRank(vvar, defBlock, defRank, useRank)
+  /**
+   * Holds if the specified `useLocation` is live on entry to `block`. This holds if there is a use of `useLocation`
+   * that is reachable from the start of `block` without passing through a definition that overlaps `useLocation`.
```

Comment:
Should this be "without passing through a definition that totally overlaps useLocation"? If not, can add a clarification about how this interacts with Chi nodes?

Reply:
Clarification added